### ISSUE TITLE
Err plugin repository and manager

### ISSUE THREAD

**Issue by swayson** [errbotio/errbot#184]
I am quite fond of Sublime Text. They have an incredible community for package development. Largely due to the Package Control repository. 

Now imagine having this for Err. Where we all can share plugins very easily and from a unified store front. This could dramatically increase Err's reach and usability.

Ref: https://sublime.wbond.net/

**Comment by gbin** [comment]
I have made a small POC around making err scan github and find its contrib plugin by itself.
https://github.com/gbin/err-catalog
Imagine it with stats, search, categories etc...

What do you think ?

**Comment by mrshu** [comment]
@gbin why not utilize PyPI for this?

Every plugin would be its own package, that could be installable via pip. Thanks to PyPI, we would get stats, search, caregories (via keywords) for free.

What I am not sure though is how (and actually if) this would work in combination with Yapsy. If just installing a module using pip would not work we could agree on a magic "template" that would be in the description of the module and would contain a "link" to its git repository that could be used to `!repos install` it. Having the plugins on PyPI would also simplify the situation with dependencies: you would no longer have to install them manually from `requirements.txt`. 

If this would work we might easily get to a similar state as Hubot is with npm (https://www.npmjs.com/search?q=hubot-).

**Comment by benhoff** [comment]
@mrshu isn't this already built in?
https://github.com/gbin/err/blob/master/errbot/plugin_manager.py#L45
https://github.com/gbin/err/blob/master/errbot/core_plugins/plugins.py#L21

**Comment by mrshu** [comment]
@benhoff the functions you point out in your comment might indeed be able to instal a python package and also a git repository in which a plugin is located it does not seem that this would work if the git repository was packaged as a python package which is what I am trying to suggest here.

The problem with that approach is that the plugins would not be stored in BOT_DATA_DIR, but in current site-packages and therefore yapsy would have to look there for plugins too. Since this directory is not that [difficult to find](https://stackoverflow.com/questions/122327/how-do-i-find-the-location-of-my-python-site-packages-directory) this might be actually doable even in the current setup.

So to answer your question @benhoff no, I do not think this is already built in =).

**Comment by gbin** [comment]
Indeed this might be a good idea, the only thing I am concerned about is the complexity linked to distutils. Lita for example does that with Ruby and it transforms a 20ish line plugin to a hundredish one.

Maybe if we could make a tool that gets a plugin as is, generates setup.py with the metadata from the .plug and requirements.txt and push it to pypi directly ?

As for the discovery, we can hardcode it by convention in errbot looking for every "err-*" root packages ?

**Comment by mrshu** [comment]
> Maybe if we could make a tool that gets a plugin as is, generates setup.py with the metadata from the .plug and requirements.txt and push it to pypi directly ?

This should not be a big problem, it might even be part of the `err` distribution itself.

> As for the discovery, we can hardcode it by convention in errbot looking for every "err-*" root packages ?

I would personaly let people choose their names but require a mention of `err plugin` in [`keywords`](https://python-packaging-user-guide.readthedocs.org/en/latest/distributing/#keywords) but I think "err-*" works too.

**Comment by benhoff** [comment]
Ah, bummer! I'm looking to do this for my own personal use so I guess it was wishful thinking on my part. 

I'm forking yapsy currently. Doesn't meet my needs and I tried digging in the source code, found daemons, and I'm moving too fast to fix upstream.
The fork's API is going to be heavily based on `BotPluginManager` from this lib as I'm much too lazy to design my own APIs :stuck_out_tongue:
Once I'm finished with my fork, my plan is to exactly what you're talking about using PyPI.

**Comment by gbin** [comment]
Any update on this ?

**Comment by benhoff** [comment]
Sorta.... I finished forking yapsy into my own plugin manager. While I understand what the original author of yapsy was trying to accomplish, we ended up not seeing eye-to-eye on implementation details.

Namely, I see almost no value in `.plug` files, so my fork doesn't emphasize it. I instead use Abstract Base Classes and inheritance to figure out what is and isn't a plugin and instantiate plugins directly instead of handing back plugin info instances. The logic for backwards compatibility with yapsy is present in my fork but unintegrated. There's not much incentive to be proactive on that front. And since I changed the repo name to `pluginmanager`, it would require a hack to be truly backwards compatible. 

I'm not sure it'd be worth it for errbot to switch off of yapsy at this point or possibly any future point. But if you ever get to the point where you end up extending the plugin logic, I can guarantee that my fork will be easier to understand. There are some serious daemons in yapsy's source.
Anyways, here's the fork: https://github.com/benhoff/pluginmanager

Are you still planning on making a tool to look for and push plugins to pypi? I think sublime (the original inspiration) requires developers to register plugins personally. There's probably more incentive to do that, as plugin downloads and such are directly tracked on the website.

**Comment by asfaltboy** [comment]
Sublime's `Package Control` asks devs to submit a PR to their "big list of plugins" repo, which includes a change to a json file with a link to the plugin zip or a link to source repo.
Their plan was to keep it simple and "stay in control" (no pun intended) of the packages listing.

I prefer the idea of using pypi since it's the "de-facto" python package registry, but there's the issue of making it easy for us to look for plugins and for plugin devs to make a setup.py, register with pypi and upload. Alternatively to using pypi.python.org, we can use our own [pypiserver](https://pypi.python.org/pypi/pypiserver) or [devpi](https://pypi.python.org/pypi/devpi) index server.

Regardless of the choice for errbot, I suggest making a `create new plugin` command-line option. A plugin template can be used to auto-generate a sane directory structure and the required packaging files ([cookiecutter](https://github.com/audreyr/cookiecutter) makes it dead simple).

**Comment by mrshu** [comment]
Thanks for your input @asfaltboy, I believe @zoni's [err-skeleton](https://github.com/zoni/err-skeleton) might be a good starting point for that. A `.plug` file filled with metadata is already required, adding a simple `setup.py` should not be a big problem.

I still believe that PyPI would be the best option for errbot, given that it is already pretty established in the Python community. Unfortunately, creating a package is not as simple as it really should be in my view. Using our own pypiserver or devpi server would not simplify this state and I am afraid if we try to simplify the process we might end up recreating PyPI.

**Comment by benhoff** [comment]
Did you guys decide that entry points were out of the question? Having dived again into python packaging, I believe that's going to be the most bang for your buck. Having seen some traffic on `repos` it would appear that you might not have gone that way, but that's the way I'm pushing in my massively plugin overloaded projects.

**Comment by zoni** [comment]
> Did you guys decide that entry points were out of the question? Having dived again into python packaging, I believe that's going to be the most bang for your buck

It's not happening for now, but it's not out of the question entirely. Basically there's some major social as well as technical issues that would need to be addressed if we were to make such a change.

Python packaging has a very steep learning curve and would be a **very high barrier of entry** to people not already very familiar with Python and its ecosystem. Unless we address this need and take away all the pitfalls that come with it, we will potentially alienate a lot of our current userbase. 

One of errbot's core strengths is that it is so **easy** to make a plugin, it takes almost no technical know-how. Even your non-technical boyfriend/girlfriend/company's marketing department/etc can do it. This is something I'm personally not willing to compromise (much) on, I strongly believe that the low barrier of entry to contribute is a major factor to adoption within a company/project.

Some of this can be addressed with tooling. We could create a script which asks some details (plugin name, description, etc) and spits out a valid `setup.py` which takes away some of the necessity to know about python packaging, but this alone is not enough.

Right now you can publish a plugin by publishing it as a git repository somewhere. With Python packaging the platform for distribution is PyPI, which requires creating an account on PyPI. Then there's the uploading to it, with `sdist` and `bdist` and wheels and whatnot.

This is improving with standardization around wheels and we could, again, making some tooling to make version bumping and PyPI uploading simple but there's only so much time in a day. :smile: 

Lastly, if we did make such a change, we would need to think about whether to keep yapsy at all. [entry_points](https://pythonhosted.org/setuptools/pkg_resources.html#entry-points) would already let us discover all installed plugins, so technically yapsy could just go. That would be a massive architectural change though. Or we could just keep yapsy and feed it a list of directories to crawl based on the installed plugins we found. Either way, we'd need to decide a strategy.

Summing up, I'm not against it, but there's so just many things to take into account both to make it work on a technical level as well as to keep it all as user-friendly as it currently is that I have little motivation to get started on it (it'd be different if I got paid to do it, but alas, I can largely only work on errbot in my spare time).

For the short term, I see more value in making the current system slightly more flexible. Being able to change the checked-out branch of a plugin would be quite simple to implement but could already do a lot for people (tagging versions, pushing a development branch for people to try out, etc etc.) that the current "we force you to use the master branch of installed plugins" makes more difficult.

**Comment by gbin** [comment]
This issue became kind of stale, please open another one name: map errbot plugins in pypi (or any other precise proposal) with a spec on how errbot discover the plugins, how users install them, how they are generated from legacy plugins (if needed) etc...
